File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-fromgeneral.rb
# LeetCode #94 - Binary Tree Inorder Traversal
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# Special case of a general recursive solution.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  out = []
  traverse(root, ->(_) { }, ->(val) { out << val }, ->(_) { })
  out
end

def traverse(root, f_pre, f_in, f_post)
  dfs = ->(node) do
    return unless node

    f_pre.call(node.val)
    dfs.call(node.left)
    f_in.call(node.val)
    dfs.call(node.right)
    f_post.call(node.val)
  end

  dfs.call(root)
end

File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-fromgeneral.java
// LeetCode #94 - Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/
// Implementation from general recursive DFS.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        dfs(root, Solution::noop, values::add, Solution::noop);
        return values;
    }

    private static void noop(int _value) {
        // Do nothing.
    }

    private static void dfs(TreeNode root,
                            IntConsumer preorderAction,
                            IntConsumer inorderAction,
                            IntConsumer postorderAction) {
        if (root == null) return;

        preorderAction.accept(root.val);
        dfs(root.left, preorderAction, inorderAction, postorderAction);
        inorderAction.accept(root.val);
        dfs(root.right, preorderAction, inorderAction, postorderAction);
        postorderAction.accept(root.val);
    }
}
